ghproxy/ghproxy.go

```diff
@@ -188,7 +191,7 @@ func main() {
 	health := pjutil.NewHealthOnPort(o.instrumentationOptions.HealthPort)
 	health.ServeReady()
 
-	interrupts.ListenAndServe(server, 30*time.Second)
+	interrupts.ListenAndServe(server, time.Duration(o.timeout)*time.Second)
```

Comment:
I have the feeling this timeout is related to the proxy timeout I am trying to change. Maybe it's bad feeling. If so, please comment and I will revert this line.

Reply:
Why did this result in the timeouts yesterday for openshift but not for k8s? Or is this change orthogonal to that issue?

Reply:
Nope, it's not directly related. We detected we can't increase v4 API throttle time due to a fact that paged queries are detected as one, and artificially timeouted when throttled.
Yesterday's problem is a lack of queue canceling in case of huge delays, which has to be introduced separately and it's not related to this effort.

Reply:
What I mean is a fix to recover from yesterday's situation when additional throttling is used will come later on.

Reply:
Could you point me to where/how paginated v4 queries are detected as one? We send multiple http requests in that case that only differ by a cursor var (that ends up being part of the http body), I am not sure I understand how ghproxy recognizes them as being part of the same operation: https://github.com/kubernetes/test-infra/blob/b1b67055258cc89f7f2ed61f88687ed4958da42f/prow/tide/search.go#L65-L80

Reply:
I have not analyzed code so deeply but I have strong observations proving there is a problem (I thought in paging, but maybe somewhere else).
Consider the delay between `v4 POST` messages to be 300 ms as implemented currently in our openshift config. There are around 90 v4 requests coming from tide so it makes around ~25-27 seconds to finish the query. If we increased throttle time to 400-500 ms we got this in tide during every sync exec:
```
tide     Failed to execute query.    {"component":"tide","controller":"sync","error":"non-200 OK status code: 503 Service Unavailable body: \"ghproxy timed out after 30s\"",

```

The same with gh apps since `openshift` org is big and it's the only org causing the problem.

If it's not the paging it could be anything else - maybe all requests are sent exactly at the same time (90 requests huh?), there are scheduled and waiting so long. The point is we have like 2 mins of sync time, we can wait on messages coming later than 30 seconds. There is no need to have a fixed limit in that case.

Reply:
I looked better into what is happening and it is indeed that all queries configured in tide get submitted in parallel at the same time, one for each iteration of the top-level `for` cycle here:

https://github.com/kubernetes/test-infra/blob/b1b67055258cc89f7f2ed61f88687ed4958da42f/prow/tide/tide.go#L454-L467

So they all hit ghproxy around the same time, ghproxy will serialize them (with delays), some of the unlucky ones that happen to be scheduled at the end of the queue go beyond the 30s mark and ghproxy will sever them.

Here's the snippet of the log that will end up the `ghproxy timed out after 30s` thing Jakub mentions, if the `duration` goes above 30s - just for illustration where I did see this happen.
[tide.log](https://github.com/kubernetes/test-infra/files/8304687/tide.log)

Reply:
It seems the real issue is that that timeout shouldn't be including the internal throttling time but rather just be a timeout for the request made to github

Reply:
Could be, but that is possibly much bigger refactoring (these are in completely different places as far as I remember).

Reply:
Not really, just give the request we make to github a `context.WithTimeout`